<a href="https://cognitiveclass.ai"><img src="https://cf-courses-data.s3.us.cloud-object-storage.appdomain.cloud/IBMDeveloperSkillsNetwork-DL0101EN-SkillsNetwork/images/IDSN-logo.png" width="400"> </a>

# Regression Models with Keras

Estimated time needed: **45** mins


## Introduction


As we discussed in the videos, despite the popularity of more powerful libraries such as PyToch and TensorFlow, they are not easy to use and have a steep learning curve. So, for people who are just starting to learn deep learning, there is no better library to use other than the Keras library. 

Keras is a high-level API for building deep learning models. It has gained favor for its ease of use and syntactic simplicity facilitating fast development. As you will see in this lab and the other labs in this course, building a very complex deep learning network can be achieved with Keras with only few lines of code. You will appreciate Keras even more, once you learn how to build deep models using PyTorch and TensorFlow in the other courses.

So, in this lab, you will learn how to use the Keras library to build a regression model.

앞서 영상에서 설명했듯이, PyTorch나 TensorFlow와 같은 강력한 라이브러리들이 인기를 얻고 있지만, 사용하기 쉽지 않고 학습 곡선이 가파릅니다. 따라서 딥러닝을 처음 배우는 사람들에게는 Keras 라이브러리보다 더 좋은 선택은 없습니다.

Keras는 딥러닝 모델 구축을 위한 고수준 API입니다. 사용 편의성과 간결한 구문 덕분에 빠른 개발이 가능하여 많은 사랑을 받고 있습니다. 이 실습과 이 강좌의 다른 실습에서 보겠지만, Keras를 사용하면 단 몇 줄의 코드로 매우 복잡한 딥러닝 네트워크를 구축할 수 있습니다. 다른 강좌에서 PyTorch와 TensorFlow를 사용하여 딥 모델을 구축하는 방법을 배우고 나면 Keras의 장점을 더욱 실감하게 될 것입니다.

이번 실습에서는 Keras 라이브러리를 사용하여 회귀 모델을 구축하는 방법을 배우게 됩니다.

## Objectives for this Notebook    
* How to use the Keras library to build a regression model
* Download and clean the data set
* Build a neural network
* Train and test the network     

## 이 노트북의 목표
* Keras 라이브러리를 사용하여 회귀 모델을 구축하는 방법
* 데이터셋 다운로드 및 정리
* 신경망 구축
* 신경망 학습 및 테스트


<h2>Table of Contents</h2>


<div class="alert alert-block alert-info" style="margin-top: 20px">

<font size = 4>
1. <a href="#Download-and-Clean-the-Data-Set">Download and Clean the Data Set</a><br>
2. <a href="#Import-Keras-Packages">Import Keras Packages</a><br>
3. <a href="#Build-a-Neural-Network">Build a Neural Network</a><br>
4. <a href="#Train-and-Test-the-Network">Train and Test the Network</a><br>  

</font>
</div>


Let's start by importing the <em>pandas</em> and the Numpy libraries.


In [1]:
# All Libraries required for this lab are listed below. 

!pip install numpy==2.0.2
!pip install pandas==2.2.2
!pip install tensorflow_cpu==2.18.0


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 19.2/19.2 MB 97.4 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 12.7/12.7 MB 67.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 230.3/230.3 MB 28.8 MB/s eta 0:00:0000:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.6/6.6 MB 35.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.1/5.1 MB 32.5 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.5/1.5 MB 52.6 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 24.5/24.5 MB 39.9 MB/s eta 0:00:0000:0100:01
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 2.2/2.2 MB 50.8 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 5.5/5.5 MB 39.0 MB/s eta 0:00:00
   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 6.6/6.6 MB 101.5 MB/s eta 0:00:00


#### To use Keras, you will also need to install a backend framework – such as TensorFlow.

If you install TensorFlow 2.16 or above, it will install Keras by default.

We are using the CPU version of tensorflow since we are dealing with smaller datasets. 
You may install the GPU version of tensorflow on your machine to accelarate the processing of larger datasets

#### Keras를 사용하려면 TensorFlow와 같은 백엔드 프레임워크도 설치해야 합니다.

TensorFlow 2.16 이상 버전을 설치하면 Keras가 기본적으로 설치됩니다.

본 예제에서는 소규모 데이터셋을 다루기 때문에 TensorFlow의 CPU 버전을 사용합니다.

더 큰 규모의 데이터셋 처리를 가속화하려면 TensorFlow의 GPU 버전을 설치할 수 있습니다.


#### Suppress the tensorflow warning messages
We use the following code to  suppress the warning messages due to use of CPU architechture for tensoflow.

You may want to **comment out** these lines if you are using the GPU architechture

#### TensorFlow 경고 메시지 억제
TensorFlow에서 CPU 아키텍처를 사용하기 때문에 발생하는 경고 메시지를 억제하기 위해 다음 코드를 사용합니다.

GPU 아키텍처를 사용하는 경우, 이 줄들을 **주석 처리**하는 것이 좋습니다.

In [2]:
import os
os.environ['TF_ENABLE_ONEDNN_OPTS'] = '0'
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '2'

In [3]:
import pandas as pd
import numpy as np
import keras

import warnings
warnings.simplefilter('ignore', FutureWarning)

/opt/conda/lib/python3.12/site-packages/keras/src/export/tf2onnx_lib.py:8: FutureWarning: In the future `np.object` will be defined as the corresponding NumPy scalar.
  if not hasattr(np, "object"):


We will be playing around with the same dataset that we used in the videos.

<strong>The dataset is about the compressive strength of different samples of concrete based on the volumes of the different ingredients that were used to make them. Ingredients include:</strong>

* Cement
* Blast furnace slag
* Fly ash
* Water
* Superplasticizer
* Coarse aggregate
* Fine aggregate
* 
우리는 영상에서 사용했던 것과 동일한 데이터셋을 가지고 실험을 진행할 것입니다.

<strong>이 데이터셋은 콘크리트를 만드는 데 사용된 다양한 재료의 부피에 따른 여러 콘크리트 시료의 압축 강도에 관한 것입니다. 재료는 다음과 같습니다.</strong>

* 시멘트
* 고로 슬래그
* 플라이애시
* 물
* 고성능 감수제
* 굵은 골재
* 잔골재

## Download and Clean the Data Set


Let's download the data and read it into a <em>pandas</em> dataframe.


In [4]:
filepath='https://s3-api.us-geo.objectstorage.softlayer.net/cf-courses-data/CognitiveClass/DL0101EN/labs/data/concrete_data.csv'
concrete_data = pd.read_csv(filepath)

concrete_data.head()

,Cement,Blast Furnace Slag,Fly Ash,Water,Superplasticizer,Coarse Aggregate,Fine Aggregate,Age,Strength
0,540.0,0.0,0.0,162.0,2.5,1040.0,676.0,28,79.99
1,540.0,0.0,0.0,162.0,2.5,1055.0,676.0,28,61.89
2,332.5,142.5,0.0,228.0,0.0,932.0,594.0,270,40.27
3,332.5,142.5,0.0,228.0,0.0,932.0,594.0,365,41.05
4,198.6,132.4,0.0,192.0,0.0,978.4,825.5,360,44.30


So the first concrete sample has 540 cubic meter of cement, 0 cubic meter of blast furnace slag, 0 cubic meter of fly ash, 162 cubic meter of water, 2.5 cubic meter of superplaticizer, 1040 cubic meter of coarse aggregate, 676 cubic meter of fine aggregate. Such a concrete mix which is 28 days old, has a compressive strength of 79.99 MPa. 

첫 번째 콘크리트 시료는 시멘트 540 세제곱미터, 고로 슬래그 0 세제곱미터, 플라이애시 0 세제곱미터, 물 162 세제곱미터, 경화제 2.5 세제곱미터, 굵은 골재 1040 세제곱미터, 잔골재 676 세제곱미터로 구성되어 있습니다. 이 콘크리트는 28일 경과 후 압축강도가 79.99 MPa입니다.

#### Let's check how many data points we have
데이터 포인트가 몇 개인지 확인해 봅시다.

In [5]:
concrete_data.shape

(1030, 9)

So, there are approximately 1000 samples to train our model on. Because of the few samples, we have to be careful not to overfit the training data.

따라서 모델 학습에 사용할 샘플은 약 1000개입니다. 샘플 수가 적기 때문에 학습 데이터에 과적합되지 않도록 주의해야 합니다.


Let's check the dataset for any missing values.
데이터셋에 누락된 값이 있는지 확인해 보겠습니다.

In [6]:
concrete_data.describe()

,Cement,Blast Furnace Slag,Fly Ash,Water,Superplasticizer,Coarse Aggregate,Fine Aggregate,Age,Strength
count,1030.000000,1030.000000,1030.000000,1030.000000,1030.000000,1030.000000,1030.000000,1030.000000,1030.000000
mean,281.167864,73.895825,54.188350,181.567282,6.204660,972.918932,773.580485,45.662136,35.817961
std,104.506364,86.279342,63.997004,21.354219,5.973841,77.753954,80.175980,63.169912,16.705742
min,102.000000,0.000000,0.000000,121.800000,0.000000,801.000000,594.000000,1.000000,2.330000
25%,192.375000,0.000000,0.000000,164.900000,0.000000,932.000000,730.950000,7.000000,23.710000
50%,272.900000,22.000000,0.000000,185.000000,6.400000,968.000000,779.500000,28.000000,34.445000
75%,350.000000,142.950000,118.300000,192.000000,10.200000,1029.400000,824.000000,56.000000,46.135000
max,540.000000,359.400000,200.100000,247.000000,32.200000,1145.000000,992.600000,365.000000,82.600000


In [8]:
#비어있는 값 확인
concrete_data.isnull().sum()

Cement                0
Blast Furnace Slag    0
Fly Ash               0
Water                 0
Superplasticizer      0
Coarse Aggregate      0
Fine Aggregate        0
Age                   0
Strength              0
dtype: int64

데이터가 매우 깔끔하고 모델 구축에 바로 사용할 수 있을 것 같습니다.

#### Split data into predictors and target


The target variable in this problem is the concrete sample strength. Therefore, our predictors will be all the other columns.

이 문제에서 목표 변수는 콘크리트 시료 강도입니다. 따라서 예측 변수는 나머지 모든 열이 될 것입니다.

In [9]:
concrete_data_columns = concrete_data.columns

In [10]:
# predictors 강도 제외 하여 등록,  target 강도 값 만 (정답지)
predictors = concrete_data[concrete_data_columns[concrete_data_columns != 'Strength']] # all columns except Strength
target = concrete_data['Strength'] # Strength column

<a id="item2"></a>


Let's do a quick sanity check of the predictors and the target dataframes.
예측 변수와 목표 데이터프레임이 제대로 설정되었는지 간단히 확인해 보겠습니다

In [11]:
predictors.head()

,Cement,Blast Furnace Slag,Fly Ash,Water,Superplasticizer,Coarse Aggregate,Fine Aggregate,Age
0,540.0,0.0,0.0,162.0,2.5,1040.0,676.0,28
1,540.0,0.0,0.0,162.0,2.5,1055.0,676.0,28
2,332.5,142.5,0.0,228.0,0.0,932.0,594.0,270
3,332.5,142.5,0.0,228.0,0.0,932.0,594.0,365
4,198.6,132.4,0.0,192.0,0.0,978.4,825.5,360


In [12]:
target.head()

0    79.99
1    61.89
2    40.27
3    41.05
4    44.30
Name: Strength, dtype: float64

Finally, the last step is to normalize the data by substracting the mean and dividing by the standard deviation. 

마지막으로, 평균을 빼고 표준편차로 나누어 데이터를 정규화하는 것이 마지막 단계입니다.

In [13]:
# 정규화 수싱 
# predictors - predictors.mean(): 각 데이터에서 해당 컬럼의 평균을 뺍니다. 이렇게 하면 데이터의 중심이 $0$이 됩니다.
# / predictors.std(): 그 결과를 표준편차로 나눕니다. 이렇게 하면 데이터가 퍼져 있는 정도(분산)가 일정하게 맞춰집니다.
predictors_norm = (predictors - predictors.mean()) / predictors.std()
predictors_norm.head()

,Cement,Blast Furnace Slag,Fly Ash,Water,Superplasticizer,Coarse Aggregate,Fine Aggregate,Age
0,2.476712,-0.856472,-0.846733,-0.916319,-0.620147,0.862735,-1.217079,-0.279597
1,2.476712,-0.856472,-0.846733,-0.916319,-0.620147,1.055651,-1.217079,-0.279597
2,0.491187,0.795140,-0.846733,2.174405,-1.038638,-0.526262,-2.239829,3.551340
3,0.491187,0.795140,-0.846733,2.174405,-1.038638,-0.526262,-2.239829,5.055221
4,-0.790075,0.678079,-0.846733,0.488555,-1.038638,0.070492,0.647569,4.976069


Let's save the number of predictors to *n_cols* since we will need this number when building our network.


In [14]:
n_cols = predictors_norm.shape[1] # number of predictors

<a id="item1"></a>


##  Import Keras Packages

##### Let's import the rest of the packages from the Keras library that we will need to build our regression model.


## Keras 패키지 가져오기

##### 회귀 모델을 구축하는 데 필요한 나머지 Keras 패키지를 가져오겠습니다.

In [15]:
from keras.models import Sequential
from keras.layers import Dense
from keras.layers import Input

## Build a Neural Network


Let's define a function that defines our regression model for us so that we can conveniently call it to create our model.

회귀 모델을 정의하는 함수를 만들어, 필요할 때 편리하게 호출하여 모델을 생성할 수 있도록 해봅시다.

In [16]:
# define regression model
def regression_model():
    # create model
    model = Sequential()
    model.add(Input(shape=(n_cols,)))
    model.add(Dense(50, activation='relu'))
    model.add(Dense(50, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

The above function create a model that has two hidden layers, each of 50 hidden units.

위 함수는 각각 50개의 은닉 유닛을 가진 두 개의 은닉층을 포함하는 모델을 생성합니다.

## Train and Test the Network


Let's call the function now to create our model.

이제 함수를 호출하여 모델을 생성해 보겠습니다.

In [17]:
# build the model
model = regression_model()

Next, we will train and test the model at the same time using the *fit* method. We will leave out 30% of the data for validation and we will train the model for 100 epochs.

다음으로, `fit` 메서드를 사용하여 모델을 동시에 학습하고 테스트하겠습니다. 데이터의 30%는 검증용으로 남겨두고, 100 에포크 동안 모델을 학습시킬 것입니다.

In [18]:
# fit the model 학습시작
model.fit(predictors_norm, target, validation_split=0.3, epochs=100, verbose=2)

Epoch 1/100
23/23 - 1s - 53ms/step - loss: 1643.7867 - val_loss: 1150.1542
Epoch 2/100
23/23 - 0s - 7ms/step - loss: 1515.5005 - val_loss: 1025.9899
Epoch 3/100
23/23 - 0s - 7ms/step - loss: 1292.0680 - val_loss: 819.7567
Epoch 4/100
23/23 - 0s - 8ms/step - loss: 956.1341 - val_loss: 566.1074
Epoch 5/100
23/23 - 0s - 8ms/step - loss: 591.0280 - val_loss: 343.7136
Epoch 6/100
23/23 - 0s - 8ms/step - loss: 339.5241 - val_loss: 227.8185
Epoch 7/100
23/23 - 0s - 7ms/step - loss: 254.3953 - val_loss: 194.6491
Epoch 8/100
23/23 - 0s - 7ms/step - loss: 234.8766 - val_loss: 188.2235
Epoch 9/100
23/23 - 0s - 7ms/step - loss: 221.2611 - val_loss: 185.5286
Epoch 10/100
23/23 - 0s - 8ms/step - loss: 210.8227 - val_loss: 178.1671
Epoch 11/100
23/23 - 0s - 6ms/step - loss: 202.3593 - val_loss: 175.3005
Epoch 12/100
23/23 - 0s - 7ms/step - loss: 196.0915 - val_loss: 171.5662
Epoch 13/100
23/23 - 0s - 7ms/step - loss: 189.7011 - val_loss: 171.6763
Epoch 14/100
23/23 - 0s - 7ms/step - loss: 184.6079 - 

<strong>You can refer to this [link](https://keras.io/models/sequential/) to learn about other functions that you can use for prediction or evaluation.</strong>

<strong>예측 또는 평가에 사용할 수 있는 다른 기능에 대해 알아보려면 [링크](https://keras.io/models/sequential/)를 참조하세요.</strong>

Feel free to vary the following and note what impact each change has on the model's performance:

1. Increase or decreate number of neurons in hidden layers
2. Add more hidden layers
3. Increase number of epochs

다음 사항들을 자유롭게 변경해 보고, 각 변경 사항이 모델 성능에 미치는 영향을 기록해 보세요.

1. 은닉층의 뉴런 수 증가 또는 감소
2. 은닉층 추가
3. 에포크 수 증가

<h3>Practice Exercise 1</h3>


Now using the same dateset,try to recreate regression model featuring five hidden layers, each with 50 nodes and ReLU activation functions, a single output layer, optimized using the Adam optimizer.

이제 동일한 데이터셋을 사용하여, 각각 50개의 노드와 ReLU 활성화 함수를 가진 5개의 은닉층, 하나의 출력층을 특징으로 하는 회귀 모델을 Adam 최적화 알고리즘을 이용하여 재현해 보십시오.

In [22]:
# Write your code here
# define regression model
def regression_model_ex1():
    # create model
    model = Sequential()
    model.add(Input(shape=(n_cols,)))
    model.add(Dense(50, activation='relu'))
    model.add(Dense(50, activation='relu'))
    model.add(Dense(50, activation='relu'))
    model.add(Dense(50, activation='relu'))
    model.add(Dense(50, activation='relu'))
    model.add(Dense(1))
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

model_ex1 = regression_model_ex1()
model_ex1.fit(predictors_norm, target, validation_split=0.3, epochs=100, verbose=2)


Epoch 1/100
23/23 - 2s - 77ms/step - loss: 1618.7977 - val_loss: 1009.9272
Epoch 2/100
23/23 - 0s - 9ms/step - loss: 851.8482 - val_loss: 352.6949
Epoch 3/100
23/23 - 0s - 9ms/step - loss: 380.8155 - val_loss: 222.3689
Epoch 4/100
23/23 - 0s - 9ms/step - loss: 243.9307 - val_loss: 194.9511
Epoch 5/100
23/23 - 0s - 8ms/step - loss: 212.5431 - val_loss: 192.5480
Epoch 6/100
23/23 - 0s - 9ms/step - loss: 190.5640 - val_loss: 196.4058
Epoch 7/100
23/23 - 0s - 8ms/step - loss: 172.6999 - val_loss: 187.8992
Epoch 8/100
23/23 - 0s - 8ms/step - loss: 157.2294 - val_loss: 177.9468
Epoch 9/100
23/23 - 0s - 8ms/step - loss: 144.6865 - val_loss: 172.2459
Epoch 10/100
23/23 - 0s - 8ms/step - loss: 127.8600 - val_loss: 181.0207
Epoch 11/100
23/23 - 0s - 8ms/step - loss: 116.1362 - val_loss: 172.9752
Epoch 12/100
23/23 - 0s - 8ms/step - loss: 102.5368 - val_loss: 148.8842
Epoch 13/100
23/23 - 0s - 8ms/step - loss: 90.2023 - val_loss: 158.9243
Epoch 14/100
23/23 - 0s - 7ms/step - loss: 78.3780 - val_l

Double-click <b>here</b> for the solution.

<!-- Your answer is below:
def regression_model():
    input_colm = predictors_norm.shape[1] # Number of input features
    # create model
    model = Sequential()
    model.add(Input(shape=(input_colm,)))  # Set the number of input features 
    model.add(Dense(50, activation='relu'))  
    model.add(Dense(50, activation='relu'))
    model.add(Dense(50, activation='relu')) 
    model.add(Dense(50, activation='relu'))
    model.add(Dense(50, activation='relu'))  
    model.add(Dense(1))  # Output layer
    
    # compile model
    model.compile(optimizer='adam', loss='mean_squared_error')
    return model

-->


<h3>Practice Exercise 2</h3>


 Train and evaluate the model simultaneously using the fit() method by reserving 10% of the data for validation and training the model for 100 epochs

fit() 메서드를 사용하여 모델의 학습과 평가를 동시에 진행합니다. 이때 데이터의 10%는 검증용으로 남겨두고, 모델은 100 에포크 동안 학습합니다.

In [23]:
# Write your code here
model_ex1 = regression_model_ex1()
model_ex1.fit(predictors_norm, target, validation_split=0.1, epochs=100, verbose=2)



Epoch 1/100
29/29 - 2s - 70ms/step - loss: 1541.4938 - val_loss: 1004.3360
Epoch 2/100
29/29 - 0s - 7ms/step - loss: 701.6469 - val_loss: 288.5514
Epoch 3/100
29/29 - 0s - 8ms/step - loss: 235.8602 - val_loss: 190.9176
Epoch 4/100
29/29 - 0s - 8ms/step - loss: 193.4423 - val_loss: 188.1973
Epoch 5/100
29/29 - 0s - 7ms/step - loss: 183.6914 - val_loss: 186.2300
Epoch 6/100
29/29 - 0s - 7ms/step - loss: 169.8430 - val_loss: 170.7611
Epoch 7/100
29/29 - 0s - 7ms/step - loss: 157.8956 - val_loss: 156.6520
Epoch 8/100
29/29 - 0s - 7ms/step - loss: 146.2568 - val_loss: 149.5247
Epoch 9/100
29/29 - 0s - 7ms/step - loss: 134.7916 - val_loss: 137.5549
Epoch 10/100
29/29 - 0s - 7ms/step - loss: 118.7292 - val_loss: 126.5048
Epoch 11/100
29/29 - 0s - 7ms/step - loss: 104.3318 - val_loss: 112.5267
Epoch 12/100
29/29 - 0s - 7ms/step - loss: 91.2471 - val_loss: 101.2956
Epoch 13/100
29/29 - 0s - 7ms/step - loss: 80.7699 - val_loss: 79.7466
Epoch 14/100
29/29 - 0s - 7ms/step - loss: 72.7826 - val_los

Double-click <b>here</b> for the solution.

<!-- Your answer is below:
# build the model
model = regression_model()
model.fit(predictors_norm, target, validation_split=0.1, epochs=100, verbose=2)

-->


Based on the results, we notice that:

- Adding more hidden layers to the model increases its capacity to learn and represent complex relationships within the data. This allows the model to better identify, as a result, the model becomes more effective at fitting the training data and potentially improving its predictions.
- By reducing the proportion of data set aside for validation and using a larger portion for training, the model has access to more examples to learn from. This additional training data helps the model improve its understanding of the underlying trends, which can lead to better overall performance.  

분석 결과를 바탕으로 다음과 같은 사실을 확인할 수 있습니다.

- 모델에 은닉층을 추가하면 데이터 내의 복잡한 관계를 학습하고 표현하는 능력이 향상됩니다. 이를 통해 모델은 관계를 더 잘 파악할 수 있게 되어, 훈련 데이터에 더 효과적으로 적응하고 예측 성능을 향상시킬 수 있습니다.

- 검증에 사용하는 데이터 비율을 줄이고 훈련에 더 많은 데이터를 활용하면 모델은 더 많은 예제를 학습에 사용할 수 있습니다. 이러한 추가 훈련 데이터는 모델이 근본적인 추세를 더 잘 이해하도록 도와주며, 결과적으로 전반적인 성능 향상으로 이어질 수 있습니다.

### Thank you for completing this lab!

This notebook was created by [Alex Aklson](https://www.linkedin.com/in/aklson/). I hope you found this lab interesting and educational. Feel free to contact me if you have any questions!


<!--
## Change Log

|  Date (YYYY-MM-DD) |  Version | Changed By  |  Change Description |
|---|---|---|---|
| 2024-11-20  | 3.0  | Aman  |  Updated the library versions to current |
| 2020-09-21  | 2.0  | Srishti  |  Migrated Lab to Markdown and added to course repo in GitLab |



<hr>

## <h3 align="center"> © IBM Corporation. All rights reserved. <h3/>


## <h3 align="center"> &#169; IBM Corporation. All rights reserved. <h3/>

